cpp/src/io/parquet/predicate_pushdown.cpp

```diff
@@ -0,0 +1,505 @@
+          stats_idx++;
+        }
+      };
+      return {min.to_device(dtype, stream), max.to_device(dtype, stream)};
+    }
+  }
+};
+
+/**
+ * @brief Converts AST expression to StatsAST for comparing with column statistics
+ * This is used in row group filtering based on predicate.
+ * statistics min value of a column is referenced by column_index*2
+ * statistics min value of a column is referenced by column_index*2+1
+ *
+ */
+class stats_expression_converter : public ast::detail::expression_transformer {
+ public:
+  stats_expression_converter(ast::expression const& expr, size_type const& num_columns)
+    : _num_columns{num_columns}
+  {
+    expr.accept(*this);
+  }
+
+  /**
+   * @copydoc ast::detail::expression_transformer::visit(ast::literal const& )
+   */
+  std::reference_wrapper<ast::expression const> visit(ast::literal const& expr) override
+  {
+    _stats_expr = std::reference_wrapper<ast::expression const>(expr);
+    return expr;
+  }
+
+  /**
+   * @copydoc ast::detail::expression_transformer::visit(ast::column_reference const& )
+   */
+  std::reference_wrapper<ast::expression const> visit(ast::column_reference const& expr) override
+  {
+    CUDF_EXPECTS(expr.get_table_source() == ast::table_reference::LEFT,
+                 "Statistics AST supports only left table");
+    CUDF_EXPECTS(expr.get_column_index() < _num_columns,
+                 "Column index cannot be more than number of columns in the table");
+    _stats_expr = std::reference_wrapper<ast::expression const>(expr);
+    return expr;
+  }
+
+  /**
+   * @copydoc ast::detail::expression_transformer::visit(ast::column_name_reference const& )
+   */
+  std::reference_wrapper<ast::expression const> visit(
+    ast::column_name_reference const& expr) override
+  {
+    CUDF_FAIL("Column name reference is not supported in statistics AST");
+  }
+
+  /**
+   * @copydoc ast::detail::expression_transformer::visit(ast::operation const& )
+   */
+  std::reference_wrapper<ast::expression const> visit(ast::operation const& expr) override
+  {
+    using cudf::ast::ast_operator;
+    auto const operands = expr.get_operands();
+    auto op             = expr.get_operator();
+
+    if (auto* v = dynamic_cast<ast::column_reference const*>(&operands[0].get())) {
+      // First operand should be column reference, second should be literal.
+      CUDF_EXPECTS(cudf::ast::detail::ast_operator_arity(op) == 2,
+                   "Only binary operations are supported on column reference");
+      CUDF_EXPECTS(dynamic_cast<ast::literal const*>(&operands[1].get()) != nullptr,
+                   "Second operand of binary operation with column reference must be a literal");
+      v->accept(*this);
+      auto const col_index = v->get_column_index();
+      switch (op) {
+        /* transform to stats conditions. op(col, literal)
+        col1 == val --> vmin <= val && vmax >= val
+        col1 != val --> !(vmin == val && vmax == val)
```

Comment:
I think this should be range-based? It should be the opposite of `==`.
```suggestion
        col1 != val --> vmin > val || vmax < val
```

Reply:
I tried the same - `vmin > val or vmax < val`. It didn't work.

Reply:
Did it fail on certain types like floats or timestamps? I had the same though as Bradley here. Why won't this comparison work?

Reply:
It's because the point of the predicate is to filter out row groups (or ultimately pages) that cannot satisfy the predicate.  In the case of `col1 != val`, the only time you could eliminate the row group entirely would be if the row group is a single value equal to `val`, in which case `min` and `max` will both be equal to `val`.  In any other case you'd have to test each row individually for the inequality.

Reply:
Assuming that `vmin <= vmax`, I think that my suggestion is still valid -- and it's more obvious that this is the opposite of the equality case.

Reasoning: `vmin > val` implies `vmax > val`, so we are checking `vmax > val && vmax < val`. That's probably equivalent to `vmax != val`. Similarly for `vmax` --> `vmin > val && vmin < val`, then apply DeMorgan's law. Do you agree?

The case I'm worried about is NaN, because `x > NaN && x < NaN` isn't equivalent to `x != NaN`. I'm not sure which one is the right choice here from the quick glance and would need to think more about it. We should carefully assess and test the behaviors of NaN values (an entire row group of NaN, a row group containing NaNs, target value set to NaN, target value set to a non-NaN, etc.). I need to look back over the tests to see if these cases are covered.

Reply:
`vmin > val || vmax < val` will not suffice in this case, because it does not handle the case `vmin < val < vmax`.  It would have to be `vmin > val || vmin < val || vmax < val || vmax > val` to work, which is far more prolix than `vmin != val || vmax != val`.  Consider a  row group with values `{1, 2, 3}` and predicate `val != 2`.  Neither `vmin > val` nor `vmax < val` will evaluate to `true`, so the row group will not be visited.  But this row group clearly contains valuues that satisfy the predicate `val != 2`.

The NaN case does warrant more investigation. I'm not at all sure how the statistics would handle that.

Reply:
Oof. Thanks, I see it now.

Let's try to add tests that involve NaN values.

Reply:
NAN values are stored as nulls in statistics, which means, its binary blob is empty. if a binary is empty, we read that row group anyway. so, it gets read and filtered at the end too.